<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 13 Additional Community Typing | Multi-omic data science with R/Bioconductor</title>
  <meta name="description" content="Course material" />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 13 Additional Community Typing | Multi-omic data science with R/Bioconductor" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="Course material" />
  <meta name="github-repo" content="microbiome/course_2022_oulu" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 13 Additional Community Typing | Multi-omic data science with R/Bioconductor" />
  
  <meta name="twitter:description" content="Course material" />
  

<meta name="author" content="University of Oulu &amp; University of Turku" />


<meta name="date" content="2022-04-24" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="study-material.html"/>
<link rel="next" href="exercise-solutions.html"/>
<script src="libs/header-attrs-2.11.3/header-attrs.js"></script>
<script src="libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />








<script src="libs/accessible-code-block-0.0.1/empty-anchor.js"></script>
<link href="libs/anchor-sections-1.0.1/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections-1.0.1/anchor-sections.js"></script>
<script src="libs/kePrint-0.0.1/kePrint.js"></script>
<link href="libs/lightable-0.0.1/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Introduction to microbiome data science</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Overview</a>
<ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#contents-and-learning-goals"><i class="fa fa-check"></i><b>1.1</b> Contents and learning goals</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#schedule-and-organizers"><i class="fa fa-check"></i><b>1.2</b> Schedule and organizers</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#how-to-apply"><i class="fa fa-check"></i><b>1.3</b> How to apply</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="overview-1.html"><a href="overview-1.html"><i class="fa fa-check"></i><b>2</b> Overview</a>
<ul>
<li class="chapter" data-level="2.1" data-path="overview-1.html"><a href="overview-1.html#introduction"><i class="fa fa-check"></i><b>2.1</b> Introduction</a></li>
<li class="chapter" data-level="2.2" data-path="overview-1.html"><a href="overview-1.html#learning-goals"><i class="fa fa-check"></i><b>2.2</b> Learning goals</a></li>
<li class="chapter" data-level="2.3" data-path="overview-1.html"><a href="overview-1.html#schedule"><i class="fa fa-check"></i><b>2.3</b> Schedule</a></li>
<li class="chapter" data-level="2.4" data-path="overview-1.html"><a href="overview-1.html#material"><i class="fa fa-check"></i><b>2.4</b> Material</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="organizers.html"><a href="organizers.html"><i class="fa fa-check"></i><b>3</b> Organizers</a>
<ul>
<li class="chapter" data-level="3.1" data-path="organizers.html"><a href="organizers.html#acknowledgments"><i class="fa fa-check"></i><b>3.1</b> Acknowledgments</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="program.html"><a href="program.html"><i class="fa fa-check"></i><b>4</b> Program</a>
<ul>
<li class="chapter" data-level="4.1" data-path="program.html"><a href="program.html#day-1"><i class="fa fa-check"></i><b>4.1</b> Day 1</a></li>
<li class="chapter" data-level="4.2" data-path="program.html"><a href="program.html#day-2---alpha-diversity"><i class="fa fa-check"></i><b>4.2</b> Day 2 - Alpha diversity</a></li>
<li class="chapter" data-level="4.3" data-path="program.html"><a href="program.html#day-3---beta-diversity"><i class="fa fa-check"></i><b>4.3</b> Day 3 - Beta diversity</a></li>
<li class="chapter" data-level="4.4" data-path="program.html"><a href="program.html#day-4--multi-omics"><i class="fa fa-check"></i><b>4.4</b> Day 4- Multi-omics</a></li>
<li class="chapter" data-level="4.5" data-path="program.html"><a href="program.html#day-5-presentations-closing"><i class="fa fa-check"></i><b>4.5</b> Day 5 : Presentations &amp; closing</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="getting-started.html"><a href="getting-started.html"><i class="fa fa-check"></i><b>5</b> Getting started</a>
<ul>
<li class="chapter" data-level="5.1" data-path="getting-started.html"><a href="getting-started.html#checklist-before-the-course"><i class="fa fa-check"></i><b>5.1</b> Checklist (before the course)</a></li>
<li class="chapter" data-level="5.2" data-path="getting-started.html"><a href="getting-started.html#support-and-resources"><i class="fa fa-check"></i><b>5.2</b> Support and resources</a></li>
<li class="chapter" data-level="5.3" data-path="getting-started.html"><a href="getting-started.html#installing-and-loading-the-required-r-packages"><i class="fa fa-check"></i><b>5.3</b> Installing and loading the required R packages</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="reproducible-reporting-with-rmarkdown.html"><a href="reproducible-reporting-with-rmarkdown.html"><i class="fa fa-check"></i><b>6</b> Reproducible reporting with Rmarkdown</a></li>
<li class="chapter" data-level="7" data-path="importing-microbiome-data.html"><a href="importing-microbiome-data.html"><i class="fa fa-check"></i><b>7</b> Importing microbiome data</a>
<ul>
<li class="chapter" data-level="7.1" data-path="importing-microbiome-data.html"><a href="importing-microbiome-data.html#data-access"><i class="fa fa-check"></i><b>7.1</b> Data access</a></li>
<li class="chapter" data-level="7.2" data-path="importing-microbiome-data.html"><a href="importing-microbiome-data.html#importing-microbiome-data-in-r"><i class="fa fa-check"></i><b>7.2</b> Importing microbiome data in R</a></li>
<li class="chapter" data-level="7.3" data-path="importing-microbiome-data.html"><a href="importing-microbiome-data.html#example-solutions"><i class="fa fa-check"></i><b>7.3</b> Example solutions</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html"><i class="fa fa-check"></i><b>8</b> Microbiome data exploration</a>
<ul>
<li class="chapter" data-level="8.1" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#data-structure"><i class="fa fa-check"></i><b>8.1</b> Data structure</a>
<ul>
<li class="chapter" data-level="8.1.1" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#transformations"><i class="fa fa-check"></i><b>8.1.1</b> Transformations</a></li>
<li class="chapter" data-level="8.1.2" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#aggregation"><i class="fa fa-check"></i><b>8.1.2</b> Aggregation</a></li>
</ul></li>
<li class="chapter" data-level="8.2" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#visualization"><i class="fa fa-check"></i><b>8.2</b> Visualization</a></li>
<li class="chapter" data-level="8.3" data-path="microbiome-data-exploration.html"><a href="microbiome-data-exploration.html#exercises-optional"><i class="fa fa-check"></i><b>8.3</b> Exercises (optional)</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="alpha-diversity.html"><a href="alpha-diversity.html"><i class="fa fa-check"></i><b>9</b> Alpha diversity</a>
<ul>
<li class="chapter" data-level="9.1" data-path="alpha-diversity.html"><a href="alpha-diversity.html#visualization-1"><i class="fa fa-check"></i><b>9.1</b> Visualization</a></li>
<li class="chapter" data-level="9.2" data-path="alpha-diversity.html"><a href="alpha-diversity.html#statistical-testing-and-comparisons"><i class="fa fa-check"></i><b>9.2</b> Statistical testing and comparisons</a></li>
<li class="chapter" data-level="9.3" data-path="alpha-diversity.html"><a href="alpha-diversity.html#exercises"><i class="fa fa-check"></i><b>9.3</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="beta-diversity.html"><a href="beta-diversity.html"><i class="fa fa-check"></i><b>10</b> Beta diversity</a>
<ul>
<li class="chapter" data-level="10.1" data-path="beta-diversity.html"><a href="beta-diversity.html#examples-of-pcoa-with-different-settings"><i class="fa fa-check"></i><b>10.1</b> Examples of PCoA with different settings</a>
<ul>
<li class="chapter" data-level="10.1.1" data-path="beta-diversity.html"><a href="beta-diversity.html#pcoa-for-asv-level-data-with-bray-curtis"><i class="fa fa-check"></i><b>10.1.1</b> PCoA for ASV-level data with Bray-Curtis</a></li>
<li class="chapter" data-level="10.1.2" data-path="beta-diversity.html"><a href="beta-diversity.html#pcoa-for-asv-level-data-with-aitchison-distance"><i class="fa fa-check"></i><b>10.1.2</b> PCoA for ASV-level data with Aitchison distance</a></li>
<li class="chapter" data-level="10.1.3" data-path="beta-diversity.html"><a href="beta-diversity.html#pcoa-aggregated-to-phylum-level"><i class="fa fa-check"></i><b>10.1.3</b> PCoA aggregated to Phylum level</a></li>
</ul></li>
<li class="chapter" data-level="10.2" data-path="beta-diversity.html"><a href="beta-diversity.html#highlighting-external-variables"><i class="fa fa-check"></i><b>10.2</b> Highlighting external variables</a>
<ul>
<li class="chapter" data-level="10.2.1" data-path="beta-diversity.html"><a href="beta-diversity.html#discrete-grouping-variable-shown-with-colors"><i class="fa fa-check"></i><b>10.2.1</b> Discrete grouping variable shown with colors</a></li>
<li class="chapter" data-level="10.2.2" data-path="beta-diversity.html"><a href="beta-diversity.html#pcoa-plot-with-continuous-variable"><i class="fa fa-check"></i><b>10.2.2</b> PCoA plot with continuous variable</a></li>
</ul></li>
<li class="chapter" data-level="10.3" data-path="beta-diversity.html"><a href="beta-diversity.html#estimating-associations-with-an-external-variable"><i class="fa fa-check"></i><b>10.3</b> Estimating associations with an external variable</a></li>
<li class="chapter" data-level="10.4" data-path="beta-diversity.html"><a href="beta-diversity.html#community-typing"><i class="fa fa-check"></i><b>10.4</b> Community typing</a></li>
<li class="chapter" data-level="10.5" data-path="beta-diversity.html"><a href="beta-diversity.html#exercises-1"><i class="fa fa-check"></i><b>10.5</b> Exercises</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html"><i class="fa fa-check"></i><b>11</b> Differential abundance analysis</a>
<ul>
<li class="chapter" data-level="11.1" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html#wilcoxon-test"><i class="fa fa-check"></i><b>11.1</b> Wilcoxon test</a></li>
<li class="chapter" data-level="11.2" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html#ancom-bc"><i class="fa fa-check"></i><b>11.2</b> ANCOM-BC</a></li>
<li class="chapter" data-level="11.3" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html#comparison-of-the-methods"><i class="fa fa-check"></i><b>11.3</b> Comparison of the methods</a></li>
<li class="chapter" data-level="11.4" data-path="differential-abundance-analysis.html"><a href="differential-abundance-analysis.html#comparison-of-abundance"><i class="fa fa-check"></i><b>11.4</b> Comparison of abundance</a></li>
</ul></li>
<li class="chapter" data-level="12" data-path="study-material.html"><a href="study-material.html"><i class="fa fa-check"></i><b>12</b> Study material</a>
<ul>
<li class="chapter" data-level="12.1" data-path="study-material.html"><a href="study-material.html#lecture-slides"><i class="fa fa-check"></i><b>12.1</b> Lecture slides</a></li>
<li class="chapter" data-level="12.2" data-path="study-material.html"><a href="study-material.html#r-programming-resources"><i class="fa fa-check"></i><b>12.2</b> R programming resources</a></li>
<li class="chapter" data-level="12.3" data-path="study-material.html"><a href="study-material.html#resources-for-treesummarizedexperiment"><i class="fa fa-check"></i><b>12.3</b> Resources for TreeSummarizedExperiment</a></li>
<li class="chapter" data-level="12.4" data-path="study-material.html"><a href="study-material.html#resources-for-phyloseq"><i class="fa fa-check"></i><b>12.4</b> Resources for phyloseq</a></li>
<li class="chapter" data-level="12.5" data-path="study-material.html"><a href="study-material.html#further-reading"><i class="fa fa-check"></i><b>12.5</b> Further reading</a></li>
</ul></li>
<li class="chapter" data-level="13" data-path="additional-community-typing.html"><a href="additional-community-typing.html"><i class="fa fa-check"></i><b>13</b> Additional Community Typing</a>
<ul>
<li class="chapter" data-level="13.1" data-path="additional-community-typing.html"><a href="additional-community-typing.html#community-composition"><i class="fa fa-check"></i><b>13.1</b> Community composition</a>
<ul>
<li class="chapter" data-level="13.1.1" data-path="additional-community-typing.html"><a href="additional-community-typing.html#composition-barplot"><i class="fa fa-check"></i><b>13.1.1</b> Composition barplot</a></li>
<li class="chapter" data-level="13.1.2" data-path="additional-community-typing.html"><a href="additional-community-typing.html#composition-heatmap"><i class="fa fa-check"></i><b>13.1.2</b> Composition heatmap</a></li>
</ul></li>
<li class="chapter" data-level="13.2" data-path="additional-community-typing.html"><a href="additional-community-typing.html#cluster-into-csts"><i class="fa fa-check"></i><b>13.2</b> Cluster into CSTs</a>
<ul>
<li class="chapter" data-level="13.2.1" data-path="additional-community-typing.html"><a href="additional-community-typing.html#elbow-method"><i class="fa fa-check"></i><b>13.2.1</b> Elbow Method</a></li>
<li class="chapter" data-level="13.2.2" data-path="additional-community-typing.html"><a href="additional-community-typing.html#silhouette-method"><i class="fa fa-check"></i><b>13.2.2</b> Silhouette Method</a></li>
<li class="chapter" data-level="13.2.3" data-path="additional-community-typing.html"><a href="additional-community-typing.html#gap-statistic-method"><i class="fa fa-check"></i><b>13.2.3</b> Gap-Statistic Method</a></li>
</ul></li>
<li class="chapter" data-level="13.3" data-path="additional-community-typing.html"><a href="additional-community-typing.html#holmes-and-mcmurdie-workflow"><i class="fa fa-check"></i><b>13.3</b> Holmes and McMurdie Workflow</a>
<ul>
<li class="chapter" data-level="13.3.1" data-path="additional-community-typing.html"><a href="additional-community-typing.html#jensen-shannon-distance"><i class="fa fa-check"></i><b>13.3.1</b> Jensen-Shannon Distance</a></li>
<li class="chapter" data-level="13.3.2" data-path="additional-community-typing.html"><a href="additional-community-typing.html#nonmetric-multidimensional-scaling"><i class="fa fa-check"></i><b>13.3.2</b> NonMetric Multidimensional Scaling</a></li>
<li class="chapter" data-level="13.3.3" data-path="additional-community-typing.html"><a href="additional-community-typing.html#chi-squared-correspondence-results"><i class="fa fa-check"></i><b>13.3.3</b> Chi-Squared Correspondence Results</a></li>
<li class="chapter" data-level="13.3.4" data-path="additional-community-typing.html"><a href="additional-community-typing.html#bray-curtis-mds"><i class="fa fa-check"></i><b>13.3.4</b> Bray-Curtis MDS</a></li>
<li class="chapter" data-level="13.3.5" data-path="additional-community-typing.html"><a href="additional-community-typing.html#bray-curtis-nmds"><i class="fa fa-check"></i><b>13.3.5</b> Bray-Curtis NMDS</a></li>
<li class="chapter" data-level="13.3.6" data-path="additional-community-typing.html"><a href="additional-community-typing.html#other-distances-from-the-philentropy-package"><i class="fa fa-check"></i><b>13.3.6</b> Other distances from the Philentropy package</a></li>
<li class="chapter" data-level="13.3.7" data-path="additional-community-typing.html"><a href="additional-community-typing.html#add-a-clustering-variable"><i class="fa fa-check"></i><b>13.3.7</b> Add a Clustering Variable</a></li>
<li class="chapter" data-level="13.3.8" data-path="additional-community-typing.html"><a href="additional-community-typing.html#more-clustering"><i class="fa fa-check"></i><b>13.3.8</b> More Clustering</a></li>
</ul></li>
<li class="chapter" data-level="13.4" data-path="additional-community-typing.html"><a href="additional-community-typing.html#cst-analysis"><i class="fa fa-check"></i><b>13.4</b> CST Analysis</a></li>
<li class="chapter" data-level="13.5" data-path="additional-community-typing.html"><a href="additional-community-typing.html#session-info"><i class="fa fa-check"></i><b>13.5</b> Session Info</a></li>
<li class="chapter" data-level="13.6" data-path="additional-community-typing.html"><a href="additional-community-typing.html#bibliography"><i class="fa fa-check"></i><b>13.6</b> Bibliography</a></li>
</ul></li>
<li class="chapter" data-level="14" data-path="exercise-solutions.html"><a href="exercise-solutions.html"><i class="fa fa-check"></i><b>14</b> Exercise Solutions</a>
<ul>
<li class="chapter" data-level="14.1" data-path="exercise-solutions.html"><a href="exercise-solutions.html#section-5"><i class="fa fa-check"></i><b>14.1</b> Section 5</a></li>
<li class="chapter" data-level="14.2" data-path="exercise-solutions.html"><a href="exercise-solutions.html#section-6"><i class="fa fa-check"></i><b>14.2</b> Section 6</a></li>
<li class="chapter" data-level="14.3" data-path="exercise-solutions.html"><a href="exercise-solutions.html#section-7"><i class="fa fa-check"></i><b>14.3</b> Section 7</a></li>
<li class="chapter" data-level="14.4" data-path="exercise-solutions.html"><a href="exercise-solutions.html#section-8"><i class="fa fa-check"></i><b>14.4</b> Section 8</a></li>
</ul></li>
<li class="chapter" data-level="15" data-path="tutorial-aims.html"><a href="tutorial-aims.html"><i class="fa fa-check"></i><b>15</b> Tutorial Aims</a></li>
<li class="chapter" data-level="16" data-path="lab-1.html"><a href="lab-1.html"><i class="fa fa-check"></i><b>16</b> Lab 1</a>
<ul>
<li class="chapter" data-level="16.1" data-path="lab-1.html"><a href="lab-1.html#unsupervised-learning-feature-selection-dimension-reduction"><i class="fa fa-check"></i><b>16.1</b> unsupervised learning: Feature selection &amp; dimension reduction</a></li>
<li class="chapter" data-level="16.2" data-path="lab-1.html"><a href="lab-1.html#load-the-data"><i class="fa fa-check"></i><b>16.2</b> Load the data</a></li>
</ul></li>
<li class="chapter" data-level="17" data-path="lab-2.html"><a href="lab-2.html"><i class="fa fa-check"></i><b>17</b> Lab 2</a>
<ul>
<li class="chapter" data-level="17.1" data-path="lab-2.html"><a href="lab-2.html#unsupervised-learning-clustering-visualization"><i class="fa fa-check"></i><b>17.1</b> unsupervised learning: clustering &amp; visualization</a></li>
</ul></li>
<li class="chapter" data-level="18" data-path="lab-3.html"><a href="lab-3.html"><i class="fa fa-check"></i><b>18</b> Lab 3</a>
<ul>
<li class="chapter" data-level="18.1" data-path="lab-3.html"><a href="lab-3.html#unsupervised-learninganalysis-and-interpretation"><i class="fa fa-check"></i><b>18.1</b> unsupervised learning:Analysis and interpretation</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Multi-omic data science with R/Bioconductor</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="additional-community-typing" class="section level1" number="13">
<h1><span class="header-section-number">Chapter 13</span> Additional Community Typing</h1>
<div class="sourceCode" id="cb180"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb180-1"><a href="additional-community-typing.html#cb180-1" aria-hidden="true"></a><span class="co"># Load data</span></span>
<span id="cb180-2"><a href="additional-community-typing.html#cb180-2" aria-hidden="true"></a><span class="kw">library</span>(mia)</span>
<span id="cb180-3"><a href="additional-community-typing.html#cb180-3" aria-hidden="true"></a><span class="kw">data</span>(enterotype)</span>
<span id="cb180-4"><a href="additional-community-typing.html#cb180-4" aria-hidden="true"></a>tse &lt;-<span class="st"> </span>enterotype</span></code></pre></div>
<div id="community-composition" class="section level2" number="13.1">
<h2><span class="header-section-number">13.1</span> Community composition</h2>
<div id="composition-barplot" class="section level3" number="13.1.1">
<h3><span class="header-section-number">13.1.1</span> Composition barplot</h3>
<p>The typical way to visualise the microbiome compositions is by using a composition barplot.
In the following, we calculate the relative abundances and the top 10 taxa. The barplot is ordered by “Bacteroidetes”:</p>
<div class="sourceCode" id="cb181"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb181-1"><a href="additional-community-typing.html#cb181-1" aria-hidden="true"></a><span class="kw">library</span>(miaViz)</span>
<span id="cb181-2"><a href="additional-community-typing.html#cb181-2" aria-hidden="true"></a></span>
<span id="cb181-3"><a href="additional-community-typing.html#cb181-3" aria-hidden="true"></a><span class="co"># Remove unwanted taxon</span></span>
<span id="cb181-4"><a href="additional-community-typing.html#cb181-4" aria-hidden="true"></a>tse &lt;-<span class="st"> </span><span class="kw">subsetTaxa</span>(tse, <span class="kw">rownames</span>(tse) <span class="op">!=</span><span class="st"> &quot;-1&quot;</span>)</span>
<span id="cb181-5"><a href="additional-community-typing.html#cb181-5" aria-hidden="true"></a><span class="co"># Only consider Sanger technique</span></span>
<span id="cb181-6"><a href="additional-community-typing.html#cb181-6" aria-hidden="true"></a>mat &lt;-<span class="st"> </span><span class="kw">subsetSamples</span>(tse, <span class="kw">colData</span>(tse)<span class="op">$</span>SeqTech <span class="op">==</span><span class="st"> &quot;Sanger&quot;</span>)</span>
<span id="cb181-7"><a href="additional-community-typing.html#cb181-7" aria-hidden="true"></a></span>
<span id="cb181-8"><a href="additional-community-typing.html#cb181-8" aria-hidden="true"></a><span class="co"># Get top taxa</span></span>
<span id="cb181-9"><a href="additional-community-typing.html#cb181-9" aria-hidden="true"></a>rel_taxa &lt;-<span class="st"> </span><span class="kw">getTopTaxa</span>(tse, <span class="dt">top =</span> <span class="dv">10</span>, <span class="dt">abund_values =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb181-10"><a href="additional-community-typing.html#cb181-10" aria-hidden="true"></a></span>
<span id="cb181-11"><a href="additional-community-typing.html#cb181-11" aria-hidden="true"></a><span class="co"># Take only the top taxa</span></span>
<span id="cb181-12"><a href="additional-community-typing.html#cb181-12" aria-hidden="true"></a>mat &lt;-<span class="st"> </span><span class="kw">subsetTaxa</span>(mat, <span class="kw">is.element</span>(<span class="kw">rownames</span>(tse), rel_taxa))</span>
<span id="cb181-13"><a href="additional-community-typing.html#cb181-13" aria-hidden="true"></a></span>
<span id="cb181-14"><a href="additional-community-typing.html#cb181-14" aria-hidden="true"></a><span class="co"># Visualise composition barplot, with samples order by &quot;Bacteroides&quot;</span></span>
<span id="cb181-15"><a href="additional-community-typing.html#cb181-15" aria-hidden="true"></a><span class="kw">plotAbundance</span>(mat, <span class="dt">abund_values =</span> <span class="st">&quot;counts&quot;</span>,</span>
<span id="cb181-16"><a href="additional-community-typing.html#cb181-16" aria-hidden="true"></a>              <span class="dt">order_rank_by =</span> <span class="st">&quot;abund&quot;</span>, <span class="dt">order_sample_by =</span> <span class="st">&quot;Bacteroides&quot;</span>)</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-1-1.png" width="672" /></p>
</div>
<div id="composition-heatmap" class="section level3" number="13.1.2">
<h3><span class="header-section-number">13.1.2</span> Composition heatmap</h3>
<p>Community composition can be visualised with a heatmap. The colour of each intersection point represents the abundance of a taxon in a specific sample.</p>
<p>Here, the CLR + Z-transformed abundances are shown.</p>
<div class="sourceCode" id="cb182"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb182-1"><a href="additional-community-typing.html#cb182-1" aria-hidden="true"></a><span class="kw">library</span>(pheatmap)</span>
<span id="cb182-2"><a href="additional-community-typing.html#cb182-2" aria-hidden="true"></a><span class="kw">library</span>(grid)</span>
<span id="cb182-3"><a href="additional-community-typing.html#cb182-3" aria-hidden="true"></a><span class="kw">library</span>(RColorBrewer)</span>
<span id="cb182-4"><a href="additional-community-typing.html#cb182-4" aria-hidden="true"></a><span class="co"># CLR and Z transforms</span></span>
<span id="cb182-5"><a href="additional-community-typing.html#cb182-5" aria-hidden="true"></a>mat &lt;-<span class="st"> </span><span class="kw">transformCounts</span>(tse, <span class="dt">abund_values =</span> <span class="st">&quot;counts&quot;</span>, <span class="dt">method =</span> <span class="st">&quot;clr&quot;</span>, <span class="dt">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb182-6"><a href="additional-community-typing.html#cb182-6" aria-hidden="true"></a>mat &lt;-<span class="st"> </span><span class="kw">transformFeatures</span>(mat, <span class="dt">abund_values =</span> <span class="st">&quot;clr&quot;</span>, <span class="dt">method =</span> <span class="st">&quot;z&quot;</span>)</span>
<span id="cb182-7"><a href="additional-community-typing.html#cb182-7" aria-hidden="true"></a><span class="co"># Take only the top taxa</span></span>
<span id="cb182-8"><a href="additional-community-typing.html#cb182-8" aria-hidden="true"></a>mat &lt;-<span class="st"> </span><span class="kw">subsetTaxa</span>(mat, <span class="kw">is.element</span>(<span class="kw">rownames</span>(tse), rel_taxa))</span>
<span id="cb182-9"><a href="additional-community-typing.html#cb182-9" aria-hidden="true"></a></span>
<span id="cb182-10"><a href="additional-community-typing.html#cb182-10" aria-hidden="true"></a><span class="co"># For annotating heat map</span></span>
<span id="cb182-11"><a href="additional-community-typing.html#cb182-11" aria-hidden="true"></a>SeqTechs &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="st">&quot;SeqTech&quot;</span> =<span class="st"> </span><span class="kw">colData</span>(mat)<span class="op">$</span>SeqTech)</span>
<span id="cb182-12"><a href="additional-community-typing.html#cb182-12" aria-hidden="true"></a><span class="kw">rownames</span>(SeqTechs) &lt;-<span class="st"> </span><span class="kw">colData</span>(mat)<span class="op">$</span>Sample_ID</span>
<span id="cb182-13"><a href="additional-community-typing.html#cb182-13" aria-hidden="true"></a><span class="co"># count matrix for pheatmap</span></span>
<span id="cb182-14"><a href="additional-community-typing.html#cb182-14" aria-hidden="true"></a>mat &lt;-<span class="st"> </span><span class="kw">assays</span>(mat)<span class="op">$</span>z</span>
<span id="cb182-15"><a href="additional-community-typing.html#cb182-15" aria-hidden="true"></a></span>
<span id="cb182-16"><a href="additional-community-typing.html#cb182-16" aria-hidden="true"></a><span class="co"># Make grid for heatmap</span></span>
<span id="cb182-17"><a href="additional-community-typing.html#cb182-17" aria-hidden="true"></a>breaks &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="op">-</span><span class="dv">2</span>, <span class="dv">2</span>, <span class="dt">length.out =</span> <span class="dv">10</span>)</span>
<span id="cb182-18"><a href="additional-community-typing.html#cb182-18" aria-hidden="true"></a><span class="kw">setHook</span>(<span class="st">&quot;grid.newpage&quot;</span>, <span class="cf">function</span>() <span class="kw">pushViewport</span>(<span class="kw">viewport</span>(<span class="dt">x =</span> <span class="dv">1</span>, <span class="dt">y =</span> <span class="dv">1</span>, <span class="dt">width =</span> <span class="fl">0.9</span>, </span>
<span id="cb182-19"><a href="additional-community-typing.html#cb182-19" aria-hidden="true"></a>                                                         <span class="dt">height =</span> <span class="fl">0.9</span>, <span class="dt">name =</span> <span class="st">&quot;vp&quot;</span>, </span>
<span id="cb182-20"><a href="additional-community-typing.html#cb182-20" aria-hidden="true"></a>                                                         <span class="dt">just =</span> <span class="kw">c</span>(<span class="st">&quot;right&quot;</span>,<span class="st">&quot;top&quot;</span>))), </span>
<span id="cb182-21"><a href="additional-community-typing.html#cb182-21" aria-hidden="true"></a>        <span class="dt">action =</span> <span class="st">&quot;prepend&quot;</span>)</span>
<span id="cb182-22"><a href="additional-community-typing.html#cb182-22" aria-hidden="true"></a><span class="kw">pheatmap</span>(mat, <span class="dt">color =</span> <span class="kw">rev</span>(<span class="kw">brewer.pal</span>(<span class="dv">9</span>, <span class="st">&quot;RdBu&quot;</span>)), <span class="dt">breaks =</span> breaks, <span class="dt">main =</span> <span class="st">&quot;SeqTechs&quot;</span>, <span class="dt">treeheight_row =</span> <span class="dv">0</span>, <span class="dt">treeheight_col =</span> <span class="dv">0</span>, <span class="dt">show_colnames =</span> <span class="dv">0</span>, <span class="dt">annotation_col =</span> SeqTechs, <span class="dt">cluster_cols =</span> F)</span>
<span id="cb182-23"><a href="additional-community-typing.html#cb182-23" aria-hidden="true"></a><span class="kw">setHook</span>(<span class="st">&quot;grid.newpage&quot;</span>, <span class="ot">NULL</span>, <span class="st">&quot;replace&quot;</span>)</span>
<span id="cb182-24"><a href="additional-community-typing.html#cb182-24" aria-hidden="true"></a><span class="kw">grid.text</span>(<span class="st">&quot;Sample&quot;</span>, <span class="dt">x =</span> <span class="fl">0.39</span>, <span class="dt">y =</span> <span class="fl">-0.04</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">16</span>))</span>
<span id="cb182-25"><a href="additional-community-typing.html#cb182-25" aria-hidden="true"></a><span class="kw">grid.text</span>(<span class="st">&quot;Genus&quot;</span>, <span class="dt">x =</span> <span class="fl">-0.04</span>, <span class="dt">y =</span> <span class="fl">0.47</span>, <span class="dt">rot =</span> <span class="dv">90</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">16</span>))</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-2-1.png" width="672" /></p>
</div>
</div>
<div id="cluster-into-csts" class="section level2" number="13.2">
<h2><span class="header-section-number">13.2</span> Cluster into CSTs</h2>
<p>The burden of specifying the number of clusters falls on the researcher. To help make an informed decision, we turn to previously established methods for doing so. In this section we introduce three such methods (aside from DMM analysis) to cluster similar samples. They include the <a href="https://uc-r.github.io/kmeans_clustering">Elbow Method, Silhouette Method, and Gap Statistic Method</a>. All of them will utilise the <a href="https://uc-r.github.io/kmeans_clustering">`kmeans’</a> algorithm which essentially assigns clusters and minimises the distance within clusters (a sum of squares calculation). The default distance metric used is the Euclidean metric.</p>
<p>The scree plot allows us to see how much of the variance is captured by each dimension in the MDS ordination.</p>
<div class="sourceCode" id="cb183"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb183-1"><a href="additional-community-typing.html#cb183-1" aria-hidden="true"></a><span class="kw">library</span>(ggplot2); th &lt;-<span class="st"> </span><span class="kw">theme_bw</span>()</span>
<span id="cb183-2"><a href="additional-community-typing.html#cb183-2" aria-hidden="true"></a></span>
<span id="cb183-3"><a href="additional-community-typing.html#cb183-3" aria-hidden="true"></a><span class="co"># reset data</span></span>
<span id="cb183-4"><a href="additional-community-typing.html#cb183-4" aria-hidden="true"></a>tse &lt;-<span class="st"> </span>enterotype</span>
<span id="cb183-5"><a href="additional-community-typing.html#cb183-5" aria-hidden="true"></a></span>
<span id="cb183-6"><a href="additional-community-typing.html#cb183-6" aria-hidden="true"></a><span class="co"># MDS analysis with the first 20 dimensions</span></span>
<span id="cb183-7"><a href="additional-community-typing.html#cb183-7" aria-hidden="true"></a>tse  &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">runMDS</span>(tse, <span class="dt">FUN =</span> vegan<span class="op">::</span>vegdist, <span class="dt">method =</span> <span class="st">&quot;bray&quot;</span>, </span>
<span id="cb183-8"><a href="additional-community-typing.html#cb183-8" aria-hidden="true"></a>                       <span class="dt">name =</span> <span class="st">&quot;MDS_BC&quot;</span>, <span class="dt">exprs_values =</span> <span class="st">&quot;counts&quot;</span>, <span class="dt">ncomponents =</span> <span class="dv">20</span>)</span>
<span id="cb183-9"><a href="additional-community-typing.html#cb183-9" aria-hidden="true"></a>ord  &lt;-<span class="st"> </span><span class="kw">reducedDim</span>(tse, <span class="st">&quot;MDS_BC&quot;</span>, <span class="dt">withDimnames =</span> <span class="ot">TRUE</span>)</span>
<span id="cb183-10"><a href="additional-community-typing.html#cb183-10" aria-hidden="true"></a><span class="co"># retrieve eigenvalues</span></span>
<span id="cb183-11"><a href="additional-community-typing.html#cb183-11" aria-hidden="true"></a>eigs &lt;-<span class="st"> </span><span class="kw">attr</span>(ord, <span class="st">&quot;eig&quot;</span>)</span>
<span id="cb183-12"><a href="additional-community-typing.html#cb183-12" aria-hidden="true"></a></span>
<span id="cb183-13"><a href="additional-community-typing.html#cb183-13" aria-hidden="true"></a><span class="co"># variance in each of the axes</span></span>
<span id="cb183-14"><a href="additional-community-typing.html#cb183-14" aria-hidden="true"></a>var &lt;-<span class="st"> </span>eigs <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(eigs)</span>
<span id="cb183-15"><a href="additional-community-typing.html#cb183-15" aria-hidden="true"></a><span class="co"># first 12 values</span></span>
<span id="cb183-16"><a href="additional-community-typing.html#cb183-16" aria-hidden="true"></a>df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">x =</span> <span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">12</span>), <span class="dt">y =</span> var[<span class="dv">1</span><span class="op">:</span><span class="dv">12</span>])</span>
<span id="cb183-17"><a href="additional-community-typing.html#cb183-17" aria-hidden="true"></a></span>
<span id="cb183-18"><a href="additional-community-typing.html#cb183-18" aria-hidden="true"></a><span class="co"># create scree plot</span></span>
<span id="cb183-19"><a href="additional-community-typing.html#cb183-19" aria-hidden="true"></a>p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(df, <span class="kw">aes</span>(x, y)) <span class="op">+</span></span>
<span id="cb183-20"><a href="additional-community-typing.html#cb183-20" aria-hidden="true"></a><span class="st">     </span><span class="kw">geom_bar</span>(<span class="dt">stat=</span><span class="st">&quot;identity&quot;</span>) <span class="op">+</span></span>
<span id="cb183-21"><a href="additional-community-typing.html#cb183-21" aria-hidden="true"></a><span class="st">     </span><span class="kw">xlab</span>(<span class="st">&quot;Principal Component&quot;</span>) <span class="op">+</span></span>
<span id="cb183-22"><a href="additional-community-typing.html#cb183-22" aria-hidden="true"></a><span class="st">     </span><span class="kw">ylab</span>(<span class="st">&quot;Variance&quot;</span>) <span class="op">+</span></span>
<span id="cb183-23"><a href="additional-community-typing.html#cb183-23" aria-hidden="true"></a><span class="st">     </span><span class="kw">ggtitle</span>(<span class="st">&quot;Scree Plot&quot;</span>)</span>
<span id="cb183-24"><a href="additional-community-typing.html#cb183-24" aria-hidden="true"></a>p</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/scree-1.png" width="672" /></p>
<p>From the scree plot (above), we can see that the two principal dimensions can account for 68% of the total variation, but at least the top 4 eigenvalues are considerable. Dimensions beyond 3 and 4 may be useful so let’s try to remove the less relevant dimensions.</p>
<div class="sourceCode" id="cb184"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb184-1"><a href="additional-community-typing.html#cb184-1" aria-hidden="true"></a><span class="co"># histogram of MDS eigenvalues from the fifth dimension onward</span></span>
<span id="cb184-2"><a href="additional-community-typing.html#cb184-2" aria-hidden="true"></a>h &lt;-<span class="st"> </span><span class="kw">hist</span>(eigs[<span class="dv">5</span><span class="op">:</span><span class="kw">length</span>(eigs)], <span class="dv">100</span>)</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-3-1.png" width="672" /></p>
<div class="sourceCode" id="cb185"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb185-1"><a href="additional-community-typing.html#cb185-1" aria-hidden="true"></a><span class="kw">plot</span>(h<span class="op">$</span>mids, h<span class="op">$</span>count, <span class="dt">log =</span> <span class="st">&quot;y&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;h&quot;</span>, <span class="dt">lwd =</span> <span class="dv">10</span>, <span class="dt">lend =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-4-1.png" width="672" /></p>
<p>As you can see, dimensions 5, 6, and 7 still stand out so we will include 7 MDS dimensions.</p>
<div id="elbow-method" class="section level3" number="13.2.1">
<h3><span class="header-section-number">13.2.1</span> Elbow Method</h3>
<p>This method graphs the sum of the sum of squares for each <span class="math inline">\(k\)</span> (number of clusters), where <span class="math inline">\(k\)</span> ranges between 1 and 10. Where the bend or `elbow’ occurs is the optimal value of <span class="math inline">\(k\)</span>.</p>
<div class="sourceCode" id="cb186"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb186-1"><a href="additional-community-typing.html#cb186-1" aria-hidden="true"></a><span class="kw">library</span>(factoextra)</span>
<span id="cb186-2"><a href="additional-community-typing.html#cb186-2" aria-hidden="true"></a></span>
<span id="cb186-3"><a href="additional-community-typing.html#cb186-3" aria-hidden="true"></a><span class="co"># take only first 7 dimensions</span></span>
<span id="cb186-4"><a href="additional-community-typing.html#cb186-4" aria-hidden="true"></a>NDIM &lt;-<span class="st"> </span><span class="dv">7</span></span>
<span id="cb186-5"><a href="additional-community-typing.html#cb186-5" aria-hidden="true"></a>x    &lt;-<span class="st"> </span>ord[, <span class="dv">1</span><span class="op">:</span>NDIM]</span>
<span id="cb186-6"><a href="additional-community-typing.html#cb186-6" aria-hidden="true"></a></span>
<span id="cb186-7"><a href="additional-community-typing.html#cb186-7" aria-hidden="true"></a><span class="co"># Elbow Method</span></span>
<span id="cb186-8"><a href="additional-community-typing.html#cb186-8" aria-hidden="true"></a>factoextra<span class="op">::</span><span class="kw">fviz_nbclust</span>(x, kmeans, <span class="dt">method =</span> <span class="st">&quot;wss&quot;</span>) <span class="op">+</span></span>
<span id="cb186-9"><a href="additional-community-typing.html#cb186-9" aria-hidden="true"></a><span class="st">                         </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="dv">3</span>, <span class="dt">linetype =</span> <span class="dv">2</span>) <span class="op">+</span></span>
<span id="cb186-10"><a href="additional-community-typing.html#cb186-10" aria-hidden="true"></a><span class="st">                         </span><span class="kw">labs</span>(<span class="dt">subtitle =</span> <span class="st">&quot;Elbow Method&quot;</span>) <span class="op">+</span><span class="st"> </span>th</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/elbow-1.png" width="672" /></p>
<p>As you can see, the bend occurs at <span class="math inline">\(k=3\)</span>, implying there are three clusters present in the enterotype data.</p>
</div>
<div id="silhouette-method" class="section level3" number="13.2.2">
<h3><span class="header-section-number">13.2.2</span> Silhouette Method</h3>
<p>This method on the otherhand returns a width for each <span class="math inline">\(k\)</span>. In this case, we want the <span class="math inline">\(k\)</span> that maximises the width.</p>
<div class="sourceCode" id="cb187"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb187-1"><a href="additional-community-typing.html#cb187-1" aria-hidden="true"></a><span class="co"># Silhouette method</span></span>
<span id="cb187-2"><a href="additional-community-typing.html#cb187-2" aria-hidden="true"></a>factoextra<span class="op">::</span><span class="kw">fviz_nbclust</span>(x, kmeans, <span class="dt">method =</span> <span class="st">&quot;silhouette&quot;</span>) <span class="op">+</span></span>
<span id="cb187-3"><a href="additional-community-typing.html#cb187-3" aria-hidden="true"></a><span class="st">                         </span><span class="kw">labs</span>(<span class="dt">subtitle =</span> <span class="st">&quot;Silhouette method&quot;</span>) <span class="op">+</span><span class="st"> </span>th</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/silhouette-1.png" width="672" /></p>
<p>The graph shows the maximum occurring at <span class="math inline">\(k=2\)</span>. <span class="math inline">\(k=3\)</span> is also strongly suggested by the plot and is consistent with what we obtained from the elbow method.</p>
</div>
<div id="gap-statistic-method" class="section level3" number="13.2.3">
<h3><span class="header-section-number">13.2.3</span> Gap-Statistic Method</h3>
<p>The Gap-Statistic Method is the most complicated among the methods discussed here. In the plot produced, we want the <span class="math inline">\(k\)</span> value that maximises the output (local and global maxima), but we also want to pay attention to where the plot jumps if the maximum value doesn’t turn out to be helpful.</p>
<div class="sourceCode" id="cb188"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb188-1"><a href="additional-community-typing.html#cb188-1" aria-hidden="true"></a><span class="co"># Gap Statistic Method</span></span>
<span id="cb188-2"><a href="additional-community-typing.html#cb188-2" aria-hidden="true"></a>factoextra<span class="op">::</span><span class="kw">fviz_nbclust</span>(x, kmeans, <span class="dt">method =</span> <span class="st">&quot;gap_stat&quot;</span>, <span class="dt">nboot =</span> <span class="dv">50</span>)<span class="op">+</span></span>
<span id="cb188-3"><a href="additional-community-typing.html#cb188-3" aria-hidden="true"></a><span class="st">                         </span><span class="kw">labs</span>(<span class="dt">subtitle =</span> <span class="st">&quot;Gap Statistic Method&quot;</span>) <span class="op">+</span><span class="st"> </span>th</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/gap-statistic-1.png" width="672" /></p>
<p><span class="math inline">\(k=6\)</span> clusters, as the peak suggests, would not be very revealing since having too many clusters can cease to be revealing and so we look to the points where the graph jumps. We can see this occurs at <span class="math inline">\(k=2\)</span> and <span class="math inline">\(k=8\)</span>. The output indicates that there are at least two clusters present. Since we have previous evidence for the existence of three clusters and <span class="math inline">\(k=3\)</span> yields a higher gap statistic than <span class="math inline">\(k=2\)</span>, <span class="math inline">\(k=3\)</span> seems to be a good choice.</p>
<p>One could argue for the existence of two or three clusters as they are both good candidates. At times like these it helps to visualise the clustering in an MDS or NMDS plot.</p>
<p>Now let’s divide the subjects into their respective clusters.</p>
<div class="sourceCode" id="cb189"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb189-1"><a href="additional-community-typing.html#cb189-1" aria-hidden="true"></a><span class="kw">library</span>(cluster)</span>
<span id="cb189-2"><a href="additional-community-typing.html#cb189-2" aria-hidden="true"></a></span>
<span id="cb189-3"><a href="additional-community-typing.html#cb189-3" aria-hidden="true"></a><span class="co"># assume 3 clusters</span></span>
<span id="cb189-4"><a href="additional-community-typing.html#cb189-4" aria-hidden="true"></a>K &lt;-<span class="st"> </span><span class="dv">3</span></span>
<span id="cb189-5"><a href="additional-community-typing.html#cb189-5" aria-hidden="true"></a>x &lt;-<span class="st"> </span>ord[, <span class="dv">1</span><span class="op">:</span>NDIM]</span>
<span id="cb189-6"><a href="additional-community-typing.html#cb189-6" aria-hidden="true"></a></span>
<span id="cb189-7"><a href="additional-community-typing.html#cb189-7" aria-hidden="true"></a>clust &lt;-<span class="st"> </span><span class="kw">as.factor</span>(<span class="kw">pam</span>(x, <span class="dt">k =</span> K, <span class="dt">cluster.only =</span> T))</span>
<span id="cb189-8"><a href="additional-community-typing.html#cb189-8" aria-hidden="true"></a><span class="co"># swapping the assignment of 2 and 3 to match ravel cst enumeration</span></span>
<span id="cb189-9"><a href="additional-community-typing.html#cb189-9" aria-hidden="true"></a>clust[clust <span class="op">==</span><span class="st"> </span><span class="dv">2</span>] &lt;-<span class="st"> </span><span class="ot">NA</span></span>
<span id="cb189-10"><a href="additional-community-typing.html#cb189-10" aria-hidden="true"></a>clust[clust <span class="op">==</span><span class="st"> </span><span class="dv">3</span>] &lt;-<span class="st"> </span><span class="dv">2</span></span>
<span id="cb189-11"><a href="additional-community-typing.html#cb189-11" aria-hidden="true"></a>clust[<span class="kw">is.na</span>(clust)] &lt;-<span class="st"> </span><span class="dv">3</span></span>
<span id="cb189-12"><a href="additional-community-typing.html#cb189-12" aria-hidden="true"></a></span>
<span id="cb189-13"><a href="additional-community-typing.html#cb189-13" aria-hidden="true"></a><span class="kw">colData</span>(tse)<span class="op">$</span>CST &lt;-<span class="st"> </span>clust</span>
<span id="cb189-14"><a href="additional-community-typing.html#cb189-14" aria-hidden="true"></a>CSTs &lt;-<span class="st"> </span><span class="kw">as.character</span>(<span class="kw">seq</span>(K))</span></code></pre></div>
<p>Let’s take a look at the MDS ordination with the Bray-Curtis dissimilarity in the first four dimensions.</p>
<div class="sourceCode" id="cb190"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb190-1"><a href="additional-community-typing.html#cb190-1" aria-hidden="true"></a><span class="kw">library</span>(scater)</span>
<span id="cb190-2"><a href="additional-community-typing.html#cb190-2" aria-hidden="true"></a><span class="kw">library</span>(RColorBrewer)</span>
<span id="cb190-3"><a href="additional-community-typing.html#cb190-3" aria-hidden="true"></a><span class="kw">library</span>(cowplot)</span>
<span id="cb190-4"><a href="additional-community-typing.html#cb190-4" aria-hidden="true"></a></span>
<span id="cb190-5"><a href="additional-community-typing.html#cb190-5" aria-hidden="true"></a><span class="co"># set up colours</span></span>
<span id="cb190-6"><a href="additional-community-typing.html#cb190-6" aria-hidden="true"></a>CSTColors &lt;-<span class="st"> </span><span class="kw">brewer.pal</span>(<span class="dv">6</span>, <span class="st">&quot;Paired&quot;</span>)[<span class="kw">c</span>(<span class="dv">2</span>, <span class="dv">5</span>, <span class="dv">3</span>)] <span class="co"># Length 6 for consistency with pre-revision CST+ colouration</span></span>
<span id="cb190-7"><a href="additional-community-typing.html#cb190-7" aria-hidden="true"></a><span class="kw">names</span>(CSTColors) &lt;-<span class="st"> </span>CSTs</span>
<span id="cb190-8"><a href="additional-community-typing.html#cb190-8" aria-hidden="true"></a></span>
<span id="cb190-9"><a href="additional-community-typing.html#cb190-9" aria-hidden="true"></a>CSTColorScale &lt;-<span class="st"> </span><span class="kw">scale_colour_manual</span>(<span class="dt">name =</span> <span class="st">&quot;CST&quot;</span>, <span class="dt">values =</span> CSTColors)</span>
<span id="cb190-10"><a href="additional-community-typing.html#cb190-10" aria-hidden="true"></a>CSTFillScale &lt;-<span class="st"> </span><span class="kw">scale_fill_manual</span>(<span class="dt">name =</span> <span class="st">&quot;CST&quot;</span>, <span class="dt">values =</span> CSTColors)</span>
<span id="cb190-11"><a href="additional-community-typing.html#cb190-11" aria-hidden="true"></a></span>
<span id="cb190-12"><a href="additional-community-typing.html#cb190-12" aria-hidden="true"></a><span class="co"># plot MDS with Bray-Curtis dimensions 1 and 2</span></span>
<span id="cb190-13"><a href="additional-community-typing.html#cb190-13" aria-hidden="true"></a>p1 &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">plotReducedDim</span>(tse, <span class="st">&quot;MDS_BC&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;CST&quot;</span>, <span class="dt">point_alpha =</span> <span class="dv">1</span>, </span>
<span id="cb190-14"><a href="additional-community-typing.html#cb190-14" aria-hidden="true"></a>                             <span class="dt">percentVar =</span> var[<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>)]<span class="op">*</span><span class="dv">100</span>) <span class="op">+</span><span class="st"> </span>th <span class="op">+</span><span class="st"> </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;Ordination by Cluster&quot;</span>) <span class="op">+</span></span>
<span id="cb190-15"><a href="additional-community-typing.html#cb190-15" aria-hidden="true"></a><span class="st">                             </span><span class="kw">theme</span>(<span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>))</span>
<span id="cb190-16"><a href="additional-community-typing.html#cb190-16" aria-hidden="true"></a><span class="co"># plot MDS with Bray-Curtis dimensions 3 and 4</span></span>
<span id="cb190-17"><a href="additional-community-typing.html#cb190-17" aria-hidden="true"></a>p2 &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">plotReducedDim</span>(tse, <span class="st">&quot;MDS_BC&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;CST&quot;</span>, <span class="dt">point_alpha =</span> <span class="dv">1</span>, </span>
<span id="cb190-18"><a href="additional-community-typing.html#cb190-18" aria-hidden="true"></a>                             <span class="dt">ncomponents =</span> <span class="kw">c</span>(<span class="dv">3</span>, <span class="dv">4</span>), <span class="dt">percentVar =</span> var[<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">3</span>, <span class="dv">4</span>)]<span class="op">*</span><span class="dv">100</span>) <span class="op">+</span><span class="st"> </span>th</span>
<span id="cb190-19"><a href="additional-community-typing.html#cb190-19" aria-hidden="true"></a></span>
<span id="cb190-20"><a href="additional-community-typing.html#cb190-20" aria-hidden="true"></a><span class="kw">plot_grid</span>(p1 <span class="op">+</span><span class="st"> </span>CSTColorScale, p2 <span class="op">+</span><span class="st"> </span>CSTColorScale, <span class="dt">nrow =</span> <span class="dv">2</span>)</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-5-1.png" width="672" /></p>
<p>And now nMDS.</p>
<div class="sourceCode" id="cb191"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb191-1"><a href="additional-community-typing.html#cb191-1" aria-hidden="true"></a>tse  &lt;-<span class="st"> </span><span class="kw">runNMDS</span>(tse, <span class="dt">FUN =</span> vegan<span class="op">::</span>vegdist, <span class="dt">method =</span> <span class="st">&quot;bray&quot;</span>, </span>
<span id="cb191-2"><a href="additional-community-typing.html#cb191-2" aria-hidden="true"></a>                <span class="dt">name =</span> <span class="st">&quot;NMDS_BC&quot;</span>, <span class="dt">exprs_values =</span> <span class="st">&quot;counts&quot;</span>, <span class="dt">ncomponents =</span> <span class="dv">20</span>)</span></code></pre></div>
<pre><code>## initial  value 4.717325 
## iter   5 value 2.138115
## iter  10 value 1.879881
## iter  15 value 1.787953
## iter  20 value 1.750342
## iter  25 value 1.730981
## iter  30 value 1.719617
## iter  30 value 1.717980
## iter  30 value 1.717898
## final  value 1.717898 
## converged</code></pre>
<div class="sourceCode" id="cb193"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb193-1"><a href="additional-community-typing.html#cb193-1" aria-hidden="true"></a>scater<span class="op">::</span><span class="kw">plotReducedDim</span>(tse, <span class="st">&quot;NMDS_BC&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;CST&quot;</span>, <span class="dt">point_alpha =</span> <span class="dv">1</span>) <span class="op">+</span><span class="st"> </span>th <span class="op">+</span><span class="st"> </span></span>
<span id="cb193-2"><a href="additional-community-typing.html#cb193-2" aria-hidden="true"></a><span class="st">        </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&quot;NMDS Bray-Curtis by Cluster&quot;</span>) <span class="op">+</span></span>
<span id="cb193-3"><a href="additional-community-typing.html#cb193-3" aria-hidden="true"></a><span class="st">        </span><span class="kw">theme</span>(<span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>)) <span class="op">+</span><span class="st"> </span>CSTColorScale</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-6-1.png" width="672" /></p>
<p>We can also look at a multitude of other distance metrics and subsets of the data.</p>
</div>
</div>
<div id="holmes-and-mcmurdie-workflow" class="section level2" number="13.3">
<h2><span class="header-section-number">13.3</span> Holmes and McMurdie Workflow</h2>
<p>For the following distance measures, we will only consider the sequencing technique “Sanger” from the enterotype data.</p>
<div class="sourceCode" id="cb194"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb194-1"><a href="additional-community-typing.html#cb194-1" aria-hidden="true"></a><span class="kw">dim</span>(tse)</span></code></pre></div>
<pre><code>## [1] 553 280</code></pre>
<div class="sourceCode" id="cb196"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb196-1"><a href="additional-community-typing.html#cb196-1" aria-hidden="true"></a><span class="co"># Subset data</span></span>
<span id="cb196-2"><a href="additional-community-typing.html#cb196-2" aria-hidden="true"></a>tse2 &lt;-<span class="st"> </span><span class="kw">subsetSamples</span>(tse, <span class="kw">colData</span>(tse)<span class="op">$</span>SeqTech <span class="op">==</span><span class="st"> &quot;Sanger&quot;</span>)</span>
<span id="cb196-3"><a href="additional-community-typing.html#cb196-3" aria-hidden="true"></a></span>
<span id="cb196-4"><a href="additional-community-typing.html#cb196-4" aria-hidden="true"></a><span class="co"># Change NAs to 0</span></span>
<span id="cb196-5"><a href="additional-community-typing.html#cb196-5" aria-hidden="true"></a>x &lt;-<span class="st"> </span><span class="kw">as.vector</span>(<span class="kw">colData</span>(tse2)<span class="op">$</span>Enterotype)</span>
<span id="cb196-6"><a href="additional-community-typing.html#cb196-6" aria-hidden="true"></a>x[<span class="kw">is.na</span>(x)] &lt;-<span class="st"> </span><span class="dv">0</span></span>
<span id="cb196-7"><a href="additional-community-typing.html#cb196-7" aria-hidden="true"></a>x &lt;-<span class="st"> </span><span class="kw">factor</span>(x, <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;1&quot;</span>, <span class="st">&quot;2&quot;</span>, <span class="st">&quot;3&quot;</span>, <span class="st">&quot;0&quot;</span>))</span>
<span id="cb196-8"><a href="additional-community-typing.html#cb196-8" aria-hidden="true"></a><span class="kw">colData</span>(tse2)<span class="op">$</span>Enterotype &lt;-<span class="st"> </span>x</span></code></pre></div>
<div class="sourceCode" id="cb197"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb197-1"><a href="additional-community-typing.html#cb197-1" aria-hidden="true"></a><span class="kw">dim</span>(tse2)</span></code></pre></div>
<pre><code>## [1] 553  41</code></pre>
<div class="sourceCode" id="cb199"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb199-1"><a href="additional-community-typing.html#cb199-1" aria-hidden="true"></a><span class="kw">table</span>(<span class="kw">colData</span>(tse2)<span class="op">$</span>Enterotype)</span></code></pre></div>
<pre><code>## 
##  1  2  3  0 
##  8  6 18  9</code></pre>
<div id="jensen-shannon-distance" class="section level3" number="13.3.1">
<h3><span class="header-section-number">13.3.1</span> Jensen-Shannon Distance</h3>
<div class="sourceCode" id="cb201"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb201-1"><a href="additional-community-typing.html#cb201-1" aria-hidden="true"></a><span class="kw">library</span>(philentropy)</span>
<span id="cb201-2"><a href="additional-community-typing.html#cb201-2" aria-hidden="true"></a></span>
<span id="cb201-3"><a href="additional-community-typing.html#cb201-3" aria-hidden="true"></a>custom_FUN &lt;-<span class="st"> </span><span class="cf">function</span> (x) <span class="kw">as.dist</span>(philentropy<span class="op">::</span><span class="kw">JSD</span>(x))</span>
<span id="cb201-4"><a href="additional-community-typing.html#cb201-4" aria-hidden="true"></a>tse2 &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">runMDS</span>(tse2, <span class="dt">FUN =</span> custom_FUN, <span class="dt">name =</span> <span class="st">&quot;MDS_JSD&quot;</span>, <span class="dt">exprs_values =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb201-5"><a href="additional-community-typing.html#cb201-5" aria-hidden="true"></a>scater<span class="op">::</span><span class="kw">plotReducedDim</span>(tse2, <span class="st">&quot;MDS_JSD&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;Enterotype&quot;</span>, </span>
<span id="cb201-6"><a href="additional-community-typing.html#cb201-6" aria-hidden="true"></a>                       <span class="dt">shape_by =</span> <span class="st">&quot;Enterotype&quot;</span>, <span class="dt">point_size =</span> <span class="dv">4</span>, <span class="dt">point_alpha =</span> <span class="dv">1</span>) <span class="op">+</span></span>
<span id="cb201-7"><a href="additional-community-typing.html#cb201-7" aria-hidden="true"></a><span class="st">                       </span><span class="kw">ggtitle</span>(<span class="st">&quot;Multidimensional Scaling using Jensen-Shannon Divergence&quot;</span>) <span class="op">+</span><span class="st"> </span>th</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-9-1.png" width="672" /></p>
</div>
<div id="nonmetric-multidimensional-scaling" class="section level3" number="13.3.2">
<h3><span class="header-section-number">13.3.2</span> NonMetric Multidimensional Scaling</h3>
<div class="sourceCode" id="cb202"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb202-1"><a href="additional-community-typing.html#cb202-1" aria-hidden="true"></a>tse2 &lt;-<span class="st"> </span>mia<span class="op">::</span><span class="kw">runNMDS</span>(tse2, <span class="dt">FUN =</span> custom_FUN, <span class="dt">nmdsFUN =</span> <span class="st">&quot;monoMDS&quot;</span>, <span class="dt">name =</span> <span class="st">&quot;NMDS_JSD&quot;</span>)</span>
<span id="cb202-2"><a href="additional-community-typing.html#cb202-2" aria-hidden="true"></a>scater<span class="op">::</span><span class="kw">plotReducedDim</span>(tse2, <span class="st">&quot;NMDS_JSD&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;Enterotype&quot;</span>, </span>
<span id="cb202-3"><a href="additional-community-typing.html#cb202-3" aria-hidden="true"></a>                       <span class="dt">shape_by =</span> <span class="st">&quot;Enterotype&quot;</span>, <span class="dt">point_size =</span> <span class="dv">4</span>, <span class="dt">point_alpha =</span> <span class="dv">1</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb202-4"><a href="additional-community-typing.html#cb202-4" aria-hidden="true"></a><span class="st">                       </span><span class="kw">ggtitle</span>(<span class="st">&quot;Non-Metric Multidimensional Scaling using Jensen-Shannon Divergence&quot;</span>) <span class="op">+</span><span class="st"> </span>th</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-10-1.png" width="672" /></p>
</div>
<div id="chi-squared-correspondence-results" class="section level3" number="13.3.3">
<h3><span class="header-section-number">13.3.3</span> Chi-Squared Correspondence Results</h3>
<div class="sourceCode" id="cb203"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb203-1"><a href="additional-community-typing.html#cb203-1" aria-hidden="true"></a>tse2 &lt;-<span class="st"> </span>mia<span class="op">::</span><span class="kw">runCCA</span>(tse2, <span class="dt">name =</span> <span class="st">&quot;PCA_CS&quot;</span>, <span class="dt">exprs_values =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb203-2"><a href="additional-community-typing.html#cb203-2" aria-hidden="true"></a>scater<span class="op">::</span><span class="kw">plotReducedDim</span>(tse2, <span class="st">&quot;PCA_CS&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;Enterotype&quot;</span>, </span>
<span id="cb203-3"><a href="additional-community-typing.html#cb203-3" aria-hidden="true"></a>                       <span class="dt">shape_by =</span> <span class="st">&quot;Enterotype&quot;</span>, <span class="dt">point_size =</span> <span class="dv">3</span>, <span class="dt">point_alpha =</span> <span class="dv">1</span>) <span class="op">+</span></span>
<span id="cb203-4"><a href="additional-community-typing.html#cb203-4" aria-hidden="true"></a><span class="st">                       </span><span class="kw">ggtitle</span>(<span class="st">&quot;Correspondance Analysis using Chi-Squared Distance&quot;</span>) <span class="op">+</span><span class="st"> </span>th</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-11-1.png" width="672" /></p>
</div>
<div id="bray-curtis-mds" class="section level3" number="13.3.4">
<h3><span class="header-section-number">13.3.4</span> Bray-Curtis MDS</h3>
<div class="sourceCode" id="cb204"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb204-1"><a href="additional-community-typing.html#cb204-1" aria-hidden="true"></a>tse2 &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">runMDS</span>(tse2, <span class="dt">FUN =</span> vegan<span class="op">::</span>vegdist, <span class="dt">name =</span> <span class="st">&quot;MDS_BC&quot;</span>, <span class="dt">exprs_values =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb204-2"><a href="additional-community-typing.html#cb204-2" aria-hidden="true"></a>scater<span class="op">::</span><span class="kw">plotReducedDim</span>(tse2, <span class="st">&quot;MDS_BC&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;Enterotype&quot;</span>, </span>
<span id="cb204-3"><a href="additional-community-typing.html#cb204-3" aria-hidden="true"></a>                       <span class="dt">shape_by =</span> <span class="st">&quot;Enterotype&quot;</span>, <span class="dt">point_size =</span> <span class="dv">4</span>, <span class="dt">point_alpha =</span> <span class="dv">1</span>) <span class="op">+</span></span>
<span id="cb204-4"><a href="additional-community-typing.html#cb204-4" aria-hidden="true"></a><span class="st">                       </span><span class="kw">ggtitle</span>(<span class="st">&quot;MDS using Bray-Curtis Dissimilarity&quot;</span>) <span class="op">+</span><span class="st"> </span>th</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-12-1.png" width="672" /></p>
</div>
<div id="bray-curtis-nmds" class="section level3" number="13.3.5">
<h3><span class="header-section-number">13.3.5</span> Bray-Curtis NMDS</h3>
<div class="sourceCode" id="cb205"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb205-1"><a href="additional-community-typing.html#cb205-1" aria-hidden="true"></a>tse2 &lt;-<span class="st"> </span>mia<span class="op">::</span><span class="kw">runNMDS</span>(tse2, <span class="dt">FUN =</span> vegan<span class="op">::</span>vegdist, <span class="dt">nmdsFUN =</span> <span class="st">&quot;monoMDS&quot;</span>, <span class="dt">method =</span> <span class="st">&quot;bray&quot;</span>, <span class="dt">name =</span> <span class="st">&quot;NMDS_BC&quot;</span>)</span>
<span id="cb205-2"><a href="additional-community-typing.html#cb205-2" aria-hidden="true"></a>scater<span class="op">::</span><span class="kw">plotReducedDim</span>(tse2, <span class="st">&quot;NMDS_BC&quot;</span>, <span class="dt">colour_by =</span> <span class="st">&quot;Enterotype&quot;</span>, </span>
<span id="cb205-3"><a href="additional-community-typing.html#cb205-3" aria-hidden="true"></a>                       <span class="dt">shape_by =</span> <span class="st">&quot;Enterotype&quot;</span>, <span class="dt">point_size =</span> <span class="dv">4</span>, <span class="dt">point_alpha =</span> <span class="dv">1</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb205-4"><a href="additional-community-typing.html#cb205-4" aria-hidden="true"></a><span class="st">                       </span><span class="kw">ggtitle</span>(<span class="st">&quot;nMDS using Bray-Curtis&quot;</span>) <span class="op">+</span><span class="st"> </span>th</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-13-1.png" width="672" /></p>
</div>
<div id="other-distances-from-the-philentropy-package" class="section level3" number="13.3.6">
<h3><span class="header-section-number">13.3.6</span> Other distances from the Philentropy package</h3>
<div class="sourceCode" id="cb206"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb206-1"><a href="additional-community-typing.html#cb206-1" aria-hidden="true"></a><span class="kw">library</span>(gridExtra)</span>
<span id="cb206-2"><a href="additional-community-typing.html#cb206-2" aria-hidden="true"></a><span class="co"># take distances that more or less match Holmes and McMurdie</span></span>
<span id="cb206-3"><a href="additional-community-typing.html#cb206-3" aria-hidden="true"></a>distances &lt;-<span class="st"> </span><span class="kw">getDistMethods</span>()[<span class="kw">c</span>(<span class="dv">1</span>, <span class="dv">2</span>, <span class="dv">23</span>, <span class="dv">42</span>)]</span>
<span id="cb206-4"><a href="additional-community-typing.html#cb206-4" aria-hidden="true"></a><span class="co"># variable to store results</span></span>
<span id="cb206-5"><a href="additional-community-typing.html#cb206-5" aria-hidden="true"></a>plist &lt;-<span class="st"> </span><span class="kw">vector</span>(<span class="st">&quot;list&quot;</span>, <span class="kw">length</span>(distances))</span>
<span id="cb206-6"><a href="additional-community-typing.html#cb206-6" aria-hidden="true"></a><span class="kw">names</span>(plist) &lt;-<span class="st"> </span>distances</span>
<span id="cb206-7"><a href="additional-community-typing.html#cb206-7" aria-hidden="true"></a></span>
<span id="cb206-8"><a href="additional-community-typing.html#cb206-8" aria-hidden="true"></a><span class="co"># calculate each distance measure</span></span>
<span id="cb206-9"><a href="additional-community-typing.html#cb206-9" aria-hidden="true"></a><span class="cf">for</span> (i <span class="cf">in</span> distances[<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(distances)]) {</span>
<span id="cb206-10"><a href="additional-community-typing.html#cb206-10" aria-hidden="true"></a>  custom_FUN &lt;-<span class="st"> </span><span class="cf">function</span> (x) <span class="kw">as.dist</span>(philentropy<span class="op">::</span><span class="kw">distance</span>(x, <span class="dt">method =</span> i, <span class="dt">p =</span> <span class="dv">1</span>))</span>
<span id="cb206-11"><a href="additional-community-typing.html#cb206-11" aria-hidden="true"></a>  tse2 &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">runMDS</span>(tse2, <span class="dt">FUN =</span> custom_FUN, <span class="dt">name =</span> i, <span class="dt">exprs_values =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb206-12"><a href="additional-community-typing.html#cb206-12" aria-hidden="true"></a>  <span class="co"># store result in plist</span></span>
<span id="cb206-13"><a href="additional-community-typing.html#cb206-13" aria-hidden="true"></a>  plist[[i]] &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">plotReducedDim</span>(tse2, i, <span class="dt">colour_by =</span> <span class="st">&quot;Enterotype&quot;</span>)</span>
<span id="cb206-14"><a href="additional-community-typing.html#cb206-14" aria-hidden="true"></a>}</span>
<span id="cb206-15"><a href="additional-community-typing.html#cb206-15" aria-hidden="true"></a></span>
<span id="cb206-16"><a href="additional-community-typing.html#cb206-16" aria-hidden="true"></a><span class="kw">library</span>(plyr)</span>
<span id="cb206-17"><a href="additional-community-typing.html#cb206-17" aria-hidden="true"></a></span>
<span id="cb206-18"><a href="additional-community-typing.html#cb206-18" aria-hidden="true"></a><span class="co"># plot</span></span>
<span id="cb206-19"><a href="additional-community-typing.html#cb206-19" aria-hidden="true"></a>df &lt;-<span class="st"> </span><span class="kw">ldply</span>(plist, <span class="cf">function</span>(x) x<span class="op">$</span>data)</span>
<span id="cb206-20"><a href="additional-community-typing.html#cb206-20" aria-hidden="true"></a><span class="kw">names</span>(df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;distance&quot;</span>, <span class="st">&quot;Axis.1&quot;</span>, <span class="st">&quot;Axis.2&quot;</span>, <span class="st">&quot;Enterotype&quot;</span>)</span>
<span id="cb206-21"><a href="additional-community-typing.html#cb206-21" aria-hidden="true"></a>p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(df, <span class="kw">aes</span>(Axis<span class="fl">.1</span>, Axis<span class="fl">.2</span>, <span class="dt">color =</span> Enterotype, <span class="dt">shape =</span> Enterotype)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb206-22"><a href="additional-community-typing.html#cb206-22" aria-hidden="true"></a><span class="st">     </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="fl">2.5</span>, <span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb206-23"><a href="additional-community-typing.html#cb206-23" aria-hidden="true"></a><span class="st">     </span><span class="kw">facet_wrap</span>(<span class="op">~</span>distance, <span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb206-24"><a href="additional-community-typing.html#cb206-24" aria-hidden="true"></a><span class="st">     </span><span class="kw">theme</span>(<span class="dt">strip.text =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">8</span>)) <span class="op">+</span></span>
<span id="cb206-25"><a href="additional-community-typing.html#cb206-25" aria-hidden="true"></a><span class="st">     </span><span class="kw">ggtitle</span>(<span class="st">&quot;MDS on various distance metrics for Enterotype dataset&quot;</span>) <span class="op">+</span><span class="st"> </span>th</span>
<span id="cb206-26"><a href="additional-community-typing.html#cb206-26" aria-hidden="true"></a>p</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-14-1.png" width="1344" /></p>
</div>
<div id="add-a-clustering-variable" class="section level3" number="13.3.7">
<h3><span class="header-section-number">13.3.7</span> Add a Clustering Variable</h3>
<p>It may be revealing to see the full data (with subsetting) clustered according to a colData variable.</p>
<div class="sourceCode" id="cb207"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb207-1"><a href="additional-community-typing.html#cb207-1" aria-hidden="true"></a><span class="co"># remove taxon</span></span>
<span id="cb207-2"><a href="additional-community-typing.html#cb207-2" aria-hidden="true"></a>tse3 &lt;-<span class="st"> </span><span class="kw">subsetTaxa</span>(tse, <span class="kw">rownames</span>(tse) <span class="op">!=</span><span class="st"> &quot;-1&quot;</span>)</span>
<span id="cb207-3"><a href="additional-community-typing.html#cb207-3" aria-hidden="true"></a></span>
<span id="cb207-4"><a href="additional-community-typing.html#cb207-4" aria-hidden="true"></a><span class="co"># Remove NAs</span></span>
<span id="cb207-5"><a href="additional-community-typing.html#cb207-5" aria-hidden="true"></a>x &lt;-<span class="st"> </span><span class="kw">as.vector</span>(<span class="kw">colData</span>(tse3)<span class="op">$</span>Enterotype)</span>
<span id="cb207-6"><a href="additional-community-typing.html#cb207-6" aria-hidden="true"></a>x &lt;-<span class="st"> </span><span class="kw">factor</span>(x)</span>
<span id="cb207-7"><a href="additional-community-typing.html#cb207-7" aria-hidden="true"></a><span class="kw">colData</span>(tse)<span class="op">$</span>Enterotype &lt;-<span class="st"> </span>x</span></code></pre></div>
<div class="sourceCode" id="cb208"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb208-1"><a href="additional-community-typing.html#cb208-1" aria-hidden="true"></a><span class="co"># loop over distances</span></span>
<span id="cb208-2"><a href="additional-community-typing.html#cb208-2" aria-hidden="true"></a><span class="cf">for</span> (i <span class="cf">in</span> distances[<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(distances)]) {</span>
<span id="cb208-3"><a href="additional-community-typing.html#cb208-3" aria-hidden="true"></a>  custom_FUN &lt;-<span class="st"> </span><span class="cf">function</span> (x) <span class="kw">as.dist</span>(philentropy<span class="op">::</span><span class="kw">distance</span>(x, <span class="dt">method =</span> i, <span class="dt">p =</span> <span class="dv">1</span>))</span>
<span id="cb208-4"><a href="additional-community-typing.html#cb208-4" aria-hidden="true"></a>  tse3 &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">runMDS</span>(tse3, <span class="dt">FUN =</span> custom_FUN, <span class="dt">name =</span> i, <span class="dt">exprs_values =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb208-5"><a href="additional-community-typing.html#cb208-5" aria-hidden="true"></a>  </span>
<span id="cb208-6"><a href="additional-community-typing.html#cb208-6" aria-hidden="true"></a>  plist[[i]] &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">plotReducedDim</span>(tse3, i, <span class="dt">colour_by =</span> <span class="st">&quot;SeqTech&quot;</span>, <span class="dt">shape_by =</span> <span class="st">&quot;Enterotype&quot;</span>)</span>
<span id="cb208-7"><a href="additional-community-typing.html#cb208-7" aria-hidden="true"></a>}</span>
<span id="cb208-8"><a href="additional-community-typing.html#cb208-8" aria-hidden="true"></a><span class="co"># plot</span></span>
<span id="cb208-9"><a href="additional-community-typing.html#cb208-9" aria-hidden="true"></a>df &lt;-<span class="st"> </span><span class="kw">ldply</span>(plist, <span class="cf">function</span>(x) x<span class="op">$</span>data)</span>
<span id="cb208-10"><a href="additional-community-typing.html#cb208-10" aria-hidden="true"></a><span class="kw">names</span>(df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;distance&quot;</span>, <span class="st">&quot;Axis.1&quot;</span>, <span class="st">&quot;Axis.2&quot;</span>, <span class="st">&quot;SeqTech&quot;</span>, <span class="st">&quot;Enterotype&quot;</span>)</span>
<span id="cb208-11"><a href="additional-community-typing.html#cb208-11" aria-hidden="true"></a>p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(df, <span class="kw">aes</span>(Axis<span class="fl">.1</span>, Axis<span class="fl">.2</span>, <span class="dt">color =</span> SeqTech, <span class="dt">shape =</span> Enterotype)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb208-12"><a href="additional-community-typing.html#cb208-12" aria-hidden="true"></a><span class="st">     </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="dv">2</span>, <span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span></span>
<span id="cb208-13"><a href="additional-community-typing.html#cb208-13" aria-hidden="true"></a><span class="st">     </span><span class="kw">facet_wrap</span>(<span class="op">~</span>distance, <span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb208-14"><a href="additional-community-typing.html#cb208-14" aria-hidden="true"></a><span class="st">     </span><span class="kw">theme</span>(<span class="dt">strip.text =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">8</span>)) <span class="op">+</span></span>
<span id="cb208-15"><a href="additional-community-typing.html#cb208-15" aria-hidden="true"></a><span class="st">     </span><span class="kw">ggtitle</span>(<span class="st">&quot;MDS on various distance metrics for Enterotype dataset&quot;</span>) <span class="op">+</span><span class="st"> </span>th</span>
<span id="cb208-16"><a href="additional-community-typing.html#cb208-16" aria-hidden="true"></a>p</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-16-1.png" width="1344" /></p>
</div>
<div id="more-clustering" class="section level3" number="13.3.8">
<h3><span class="header-section-number">13.3.8</span> More Clustering</h3>
<p>Dropping 9 outliers (the 0 enterotype), let’s visualise the clustering in a different manner. Based on our analysis above, we concluded there were 3 distinct clusters. The authors of the Nature paper published on the enterotype data created a figure similar to the following:</p>
<div class="sourceCode" id="cb209"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb209-1"><a href="additional-community-typing.html#cb209-1" aria-hidden="true"></a><span class="kw">library</span>(ade4)</span>
<span id="cb209-2"><a href="additional-community-typing.html#cb209-2" aria-hidden="true"></a><span class="co"># remove 9 outliers</span></span>
<span id="cb209-3"><a href="additional-community-typing.html#cb209-3" aria-hidden="true"></a>tse2 &lt;-<span class="st"> </span><span class="kw">subsetSamples</span>(tse2, <span class="kw">colData</span>(tse2)<span class="op">$</span>Enterotype <span class="op">!=</span><span class="st"> </span><span class="dv">0</span>)</span>
<span id="cb209-4"><a href="additional-community-typing.html#cb209-4" aria-hidden="true"></a><span class="kw">colData</span>(tse2)<span class="op">$</span>Enterotype &lt;-<span class="st"> </span><span class="kw">factor</span>(<span class="kw">colData</span>(tse2)<span class="op">$</span>Enterotype)</span></code></pre></div>
<div class="sourceCode" id="cb210"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb210-1"><a href="additional-community-typing.html#cb210-1" aria-hidden="true"></a>clus &lt;-<span class="st"> </span><span class="kw">colData</span>(tse2)<span class="op">$</span>Enterotype</span>
<span id="cb210-2"><a href="additional-community-typing.html#cb210-2" aria-hidden="true"></a>dist &lt;-<span class="st"> </span><span class="kw">calculateJSD</span>(tse2)</span>
<span id="cb210-3"><a href="additional-community-typing.html#cb210-3" aria-hidden="true"></a>pcoa &lt;-<span class="st"> </span><span class="kw">dudi.pco</span>(dist, <span class="dt">scannf =</span> F, <span class="dt">nf =</span> <span class="dv">3</span>)</span>
<span id="cb210-4"><a href="additional-community-typing.html#cb210-4" aria-hidden="true"></a><span class="kw">s.class</span>(pcoa<span class="op">$</span>li, <span class="dt">grid =</span> F, <span class="dt">fac =</span> clus)</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-18-1.png" width="672" /></p>
<p>Revisit the distances without the 9 outlier values.</p>
<div class="sourceCode" id="cb211"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb211-1"><a href="additional-community-typing.html#cb211-1" aria-hidden="true"></a><span class="co"># loop over distances</span></span>
<span id="cb211-2"><a href="additional-community-typing.html#cb211-2" aria-hidden="true"></a><span class="cf">for</span> (i <span class="cf">in</span> distances[<span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(distances)]) {</span>
<span id="cb211-3"><a href="additional-community-typing.html#cb211-3" aria-hidden="true"></a>  custom_FUN &lt;-<span class="st"> </span><span class="cf">function</span> (x) <span class="kw">as.dist</span>(philentropy<span class="op">::</span><span class="kw">distance</span>(x, <span class="dt">method =</span> i, <span class="dt">p =</span> <span class="dv">1</span>))</span>
<span id="cb211-4"><a href="additional-community-typing.html#cb211-4" aria-hidden="true"></a>  tse2 &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">runMDS</span>(tse2, <span class="dt">FUN =</span> custom_FUN, <span class="dt">name =</span> i, <span class="dt">exprs_values =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb211-5"><a href="additional-community-typing.html#cb211-5" aria-hidden="true"></a>  </span>
<span id="cb211-6"><a href="additional-community-typing.html#cb211-6" aria-hidden="true"></a>  plist[[i]] &lt;-<span class="st"> </span>scater<span class="op">::</span><span class="kw">plotReducedDim</span>(tse2, i, <span class="dt">colour_by =</span> <span class="st">&quot;Enterotype&quot;</span>)</span>
<span id="cb211-7"><a href="additional-community-typing.html#cb211-7" aria-hidden="true"></a>}</span>
<span id="cb211-8"><a href="additional-community-typing.html#cb211-8" aria-hidden="true"></a><span class="co"># plot</span></span>
<span id="cb211-9"><a href="additional-community-typing.html#cb211-9" aria-hidden="true"></a>df &lt;-<span class="st"> </span><span class="kw">ldply</span>(plist, <span class="cf">function</span>(x) x<span class="op">$</span>data)</span>
<span id="cb211-10"><a href="additional-community-typing.html#cb211-10" aria-hidden="true"></a><span class="kw">names</span>(df) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;distance&quot;</span>, <span class="st">&quot;Axis.1&quot;</span>, <span class="st">&quot;Axis.2&quot;</span>, <span class="st">&quot;Enterotype&quot;</span>)</span>
<span id="cb211-11"><a href="additional-community-typing.html#cb211-11" aria-hidden="true"></a>p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(df, <span class="kw">aes</span>(Axis<span class="fl">.1</span>, Axis<span class="fl">.2</span>, <span class="dt">color =</span> Enterotype, <span class="dt">shape =</span> Enterotype)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb211-12"><a href="additional-community-typing.html#cb211-12" aria-hidden="true"></a><span class="st">     </span><span class="kw">geom_point</span>(<span class="dt">size =</span> <span class="fl">2.5</span>, <span class="dt">alpha =</span> <span class="fl">0.5</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb211-13"><a href="additional-community-typing.html#cb211-13" aria-hidden="true"></a><span class="st">     </span><span class="kw">facet_wrap</span>(<span class="op">~</span>distance, <span class="dt">scales =</span> <span class="st">&quot;free&quot;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb211-14"><a href="additional-community-typing.html#cb211-14" aria-hidden="true"></a><span class="st">     </span><span class="kw">theme</span>(<span class="dt">strip.text =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">8</span>)) <span class="op">+</span></span>
<span id="cb211-15"><a href="additional-community-typing.html#cb211-15" aria-hidden="true"></a><span class="st">     </span><span class="kw">ggtitle</span>(<span class="st">&quot;MDS on various distance metrics for Enterotype dataset with 9 samples removed&quot;</span>) <span class="op">+</span><span class="st"> </span>th</span>
<span id="cb211-16"><a href="additional-community-typing.html#cb211-16" aria-hidden="true"></a>p</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-19-1.png" width="1344" /></p>
<p>Would we use the same number of clusters as before with 9 values missing?</p>
<div class="sourceCode" id="cb212"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb212-1"><a href="additional-community-typing.html#cb212-1" aria-hidden="true"></a><span class="co">## Slightly faster way to use pam</span></span>
<span id="cb212-2"><a href="additional-community-typing.html#cb212-2" aria-hidden="true"></a>pam1 &lt;-<span class="st"> </span><span class="cf">function</span>(x,k) <span class="kw">list</span>(<span class="dt">cluster =</span> <span class="kw">pam</span>(x,k, <span class="dt">cluster.only=</span><span class="ot">TRUE</span>))</span>
<span id="cb212-3"><a href="additional-community-typing.html#cb212-3" aria-hidden="true"></a></span>
<span id="cb212-4"><a href="additional-community-typing.html#cb212-4" aria-hidden="true"></a>pamPCoA =<span class="st"> </span><span class="cf">function</span>(x, k) {</span>
<span id="cb212-5"><a href="additional-community-typing.html#cb212-5" aria-hidden="true"></a>    <span class="kw">list</span>(<span class="dt">cluster =</span> <span class="kw">pam</span>(x[,<span class="dv">1</span><span class="op">:</span><span class="dv">2</span>], k, <span class="dt">cluster.only =</span> <span class="ot">TRUE</span>))</span>
<span id="cb212-6"><a href="additional-community-typing.html#cb212-6" aria-hidden="true"></a>}</span>
<span id="cb212-7"><a href="additional-community-typing.html#cb212-7" aria-hidden="true"></a><span class="co"># gap statistic method</span></span>
<span id="cb212-8"><a href="additional-community-typing.html#cb212-8" aria-hidden="true"></a>gs &lt;-<span class="st"> </span><span class="kw">clusGap</span>(<span class="kw">assays</span>(tse2)<span class="op">$</span>counts, <span class="dt">FUN =</span> pamPCoA, <span class="dt">K.max =</span> <span class="dv">10</span>)</span>
<span id="cb212-9"><a href="additional-community-typing.html#cb212-9" aria-hidden="true"></a><span class="kw">fviz_gap_stat</span>(gs)</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-20-1.png" width="672" /></p>
<p>We still see a jump at three so it is still a good choice for the number of clusters.</p>
</div>
</div>
<div id="cst-analysis" class="section level2" number="13.4">
<h2><span class="header-section-number">13.4</span> CST Analysis</h2>
<p>Looking at heat maps of the CSTs can reveal structure on a different level. Using all of the data again (for the top 20 taxa) and taking the z transformation of the clr transformed counts, we have</p>
<div class="sourceCode" id="cb213"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb213-1"><a href="additional-community-typing.html#cb213-1" aria-hidden="true"></a><span class="co"># Find top 20 taxa</span></span>
<span id="cb213-2"><a href="additional-community-typing.html#cb213-2" aria-hidden="true"></a>rel_taxa &lt;-<span class="st"> </span><span class="kw">getTopTaxa</span>(tse, <span class="dt">top =</span> <span class="dv">20</span>)</span>
<span id="cb213-3"><a href="additional-community-typing.html#cb213-3" aria-hidden="true"></a><span class="co"># Z transform of CLR counts</span></span>
<span id="cb213-4"><a href="additional-community-typing.html#cb213-4" aria-hidden="true"></a>mat &lt;-<span class="st"> </span><span class="kw">transformCounts</span>(tse, <span class="dt">abund_values =</span> <span class="st">&quot;counts&quot;</span>, <span class="dt">method =</span> <span class="st">&quot;clr&quot;</span>, <span class="dt">pseudocount =</span> <span class="dv">1</span>)</span>
<span id="cb213-5"><a href="additional-community-typing.html#cb213-5" aria-hidden="true"></a>mat &lt;-<span class="st"> </span><span class="kw">transformFeatures</span>(mat, <span class="dt">abund_values =</span> <span class="st">&quot;clr&quot;</span>, <span class="dt">method =</span> <span class="st">&quot;z&quot;</span>)</span>
<span id="cb213-6"><a href="additional-community-typing.html#cb213-6" aria-hidden="true"></a><span class="co"># Select top 20 taxa</span></span>
<span id="cb213-7"><a href="additional-community-typing.html#cb213-7" aria-hidden="true"></a>mat &lt;-<span class="st"> </span><span class="kw">subsetTaxa</span>(mat, <span class="kw">is.element</span>(<span class="kw">rownames</span>(tse), rel_taxa))</span>
<span id="cb213-8"><a href="additional-community-typing.html#cb213-8" aria-hidden="true"></a>mat &lt;-<span class="st"> </span><span class="kw">assays</span>(mat)<span class="op">$</span>z</span>
<span id="cb213-9"><a href="additional-community-typing.html#cb213-9" aria-hidden="true"></a></span>
<span id="cb213-10"><a href="additional-community-typing.html#cb213-10" aria-hidden="true"></a><span class="co"># Order CSTs</span></span>
<span id="cb213-11"><a href="additional-community-typing.html#cb213-11" aria-hidden="true"></a>mat &lt;-<span class="st"> </span>mat[, <span class="kw">order</span>(clust)]</span>
<span id="cb213-12"><a href="additional-community-typing.html#cb213-12" aria-hidden="true"></a><span class="kw">colnames</span>(mat) &lt;-<span class="st"> </span><span class="kw">names</span>(<span class="kw">sort</span>(clust))</span></code></pre></div>
<div class="sourceCode" id="cb214"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb214-1"><a href="additional-community-typing.html#cb214-1" aria-hidden="true"></a><span class="co"># Plot</span></span>
<span id="cb214-2"><a href="additional-community-typing.html#cb214-2" aria-hidden="true"></a>CSTs        &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(<span class="kw">sort</span>(clust))</span>
<span id="cb214-3"><a href="additional-community-typing.html#cb214-3" aria-hidden="true"></a><span class="kw">names</span>(CSTs) &lt;-<span class="st"> &quot;CST&quot;</span></span>
<span id="cb214-4"><a href="additional-community-typing.html#cb214-4" aria-hidden="true"></a>breaks &lt;-<span class="st"> </span><span class="kw">seq</span>(<span class="op">-</span><span class="dv">2</span>, <span class="dv">2</span>, <span class="dt">length.out =</span> <span class="dv">10</span>)</span>
<span id="cb214-5"><a href="additional-community-typing.html#cb214-5" aria-hidden="true"></a><span class="co"># Make grid for heatmap</span></span>
<span id="cb214-6"><a href="additional-community-typing.html#cb214-6" aria-hidden="true"></a><span class="kw">setHook</span>(<span class="st">&quot;grid.newpage&quot;</span>, <span class="cf">function</span>() <span class="kw">pushViewport</span>(<span class="kw">viewport</span>(<span class="dt">x =</span> <span class="dv">1</span>, <span class="dt">y =</span> <span class="dv">1</span>, <span class="dt">width =</span> <span class="fl">0.9</span>, </span>
<span id="cb214-7"><a href="additional-community-typing.html#cb214-7" aria-hidden="true"></a>                                                         <span class="dt">height =</span> <span class="fl">0.9</span>, <span class="dt">name =</span> <span class="st">&quot;vp&quot;</span>, </span>
<span id="cb214-8"><a href="additional-community-typing.html#cb214-8" aria-hidden="true"></a>                                                         <span class="dt">just =</span> <span class="kw">c</span>(<span class="st">&quot;right&quot;</span>,<span class="st">&quot;top&quot;</span>))), </span>
<span id="cb214-9"><a href="additional-community-typing.html#cb214-9" aria-hidden="true"></a>        <span class="dt">action =</span> <span class="st">&quot;prepend&quot;</span>)</span>
<span id="cb214-10"><a href="additional-community-typing.html#cb214-10" aria-hidden="true"></a><span class="kw">pheatmap</span>(mat, <span class="dt">color =</span> <span class="kw">rev</span>(<span class="kw">brewer.pal</span>(<span class="dv">9</span>, <span class="st">&quot;RdBu&quot;</span>)), <span class="dt">breaks =</span> breaks, <span class="dt">main =</span> <span class="st">&quot;All CSTs&quot;</span>, <span class="dt">treeheight_row =</span> <span class="dv">0</span>, <span class="dt">treeheight_col =</span> <span class="dv">0</span>, <span class="dt">show_colnames =</span> <span class="dv">0</span>, <span class="dt">annotation_col =</span> CSTs, <span class="dt">cluster_cols =</span> F)</span>
<span id="cb214-11"><a href="additional-community-typing.html#cb214-11" aria-hidden="true"></a><span class="kw">setHook</span>(<span class="st">&quot;grid.newpage&quot;</span>, <span class="ot">NULL</span>, <span class="st">&quot;replace&quot;</span>)</span>
<span id="cb214-12"><a href="additional-community-typing.html#cb214-12" aria-hidden="true"></a><span class="kw">grid.text</span>(<span class="st">&quot;Sample&quot;</span>, <span class="dt">x =</span> <span class="fl">0.39</span>, <span class="dt">y =</span> <span class="fl">-0.04</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">16</span>))</span>
<span id="cb214-13"><a href="additional-community-typing.html#cb214-13" aria-hidden="true"></a><span class="kw">grid.text</span>(<span class="st">&quot;Genus&quot;</span>, <span class="dt">x =</span> <span class="fl">-0.04</span>, <span class="dt">y =</span> <span class="fl">0.47</span>, <span class="dt">rot =</span> <span class="dv">90</span>, <span class="dt">gp =</span> <span class="kw">gpar</span>(<span class="dt">fontsize =</span> <span class="dv">16</span>))</span></code></pre></div>
<p><img src="10-community-typing_files/figure-html/unnamed-chunk-22-1.png" width="672" /></p>
</div>
<div id="session-info" class="section level2" number="13.5">
<h2><span class="header-section-number">13.5</span> Session Info</h2>
<div class="sourceCode" id="cb215"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb215-1"><a href="additional-community-typing.html#cb215-1" aria-hidden="true"></a><span class="kw">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.1.2 (2021-11-01)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 21.10
## 
## Matrix products: default
## BLAS:   /home/xxx/bin/R-4.1.2/lib/libRblas.so
## LAPACK: /home/xxx/bin/R-4.1.2/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] grid      stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] ade4_1.7-18                    plyr_1.8.6                    
##  [3] gridExtra_2.3                  philentropy_0.6.0             
##  [5] cowplot_1.1.1                  scater_1.22.0                 
##  [7] scuttle_1.4.0                  cluster_2.1.2                 
##  [9] factoextra_1.0.7               RColorBrewer_1.1-3            
## [11] pheatmap_1.0.12                miaViz_1.3.2                  
## [13] ggraph_2.0.5                   ggplot2_3.3.5                 
## [15] mia_1.3.18                     MultiAssayExperiment_1.20.0   
## [17] TreeSummarizedExperiment_2.2.0 Biostrings_2.62.0             
## [19] XVector_0.34.0                 SingleCellExperiment_1.16.0   
## [21] SummarizedExperiment_1.24.0    Biobase_2.54.0                
## [23] GenomicRanges_1.46.1           GenomeInfoDb_1.30.1           
## [25] IRanges_2.28.0                 S4Vectors_0.32.3              
## [27] BiocGenerics_0.40.0            MatrixGenerics_1.6.0          
## [29] matrixStats_0.61.0            
## 
## loaded via a namespace (and not attached):
##   [1] backports_1.4.1             igraph_1.2.11              
##   [3] lazyeval_0.2.2              splines_4.1.2              
##   [5] BiocParallel_1.28.3         digest_0.6.29              
##   [7] yulab.utils_0.0.4           htmltools_0.5.2            
##   [9] viridis_0.6.2               fansi_1.0.3                
##  [11] magrittr_2.0.3              memoise_2.0.1              
##  [13] ScaledMatrix_1.2.0          DECIPHER_2.22.0            
##  [15] graphlayouts_0.8.0          colorspace_2.0-3           
##  [17] blob_1.2.2                  ggrepel_0.9.1              
##  [19] xfun_0.29                   dplyr_1.0.8                
##  [21] crayon_1.5.1                RCurl_1.98-1.6             
##  [23] jsonlite_1.8.0              ape_5.6-1                  
##  [25] glue_1.6.2                  polyclip_1.10-0            
##  [27] gtable_0.3.0                zlibbioc_1.40.0            
##  [29] DelayedArray_0.20.0         car_3.0-12                 
##  [31] BiocSingular_1.10.0         abind_1.4-5                
##  [33] scales_1.2.0                DBI_1.1.2                  
##  [35] rstatix_0.7.0               Rcpp_1.0.8                 
##  [37] viridisLite_0.4.0           decontam_1.14.0            
##  [39] gridGraphics_0.5-1          tidytree_0.3.8             
##  [41] bit_4.0.4                   rsvd_1.0.5                 
##  [43] ellipsis_0.3.2              pkgconfig_2.0.3            
##  [45] farver_2.1.0                sass_0.4.0                 
##  [47] utf8_1.2.2                  ggplotify_0.1.0            
##  [49] tidyselect_1.1.1            labeling_0.4.2             
##  [51] rlang_1.0.2                 reshape2_1.4.4             
##  [53] munsell_0.5.0               tools_4.1.2                
##  [55] cachem_1.0.6                cli_3.2.0                  
##  [57] DirichletMultinomial_1.36.0 generics_0.1.2             
##  [59] RSQLite_2.2.10              broom_0.7.12               
##  [61] evaluate_0.15               stringr_1.4.0              
##  [63] fastmap_1.1.0               yaml_2.3.4                 
##  [65] ggtree_3.2.1                knitr_1.37                 
##  [67] bit64_4.0.5                 tidygraph_1.2.0            
##  [69] purrr_0.3.4                 nlme_3.1-155               
##  [71] sparseMatrixStats_1.6.0     aplot_0.1.2                
##  [73] compiler_4.1.2              beeswarm_0.4.0             
##  [75] ggsignif_0.6.3              treeio_1.18.1              
##  [77] tibble_3.1.6                tweenr_1.0.2               
##  [79] bslib_0.3.1                 stringi_1.7.6              
##  [81] highr_0.9                   lattice_0.20-45            
##  [83] Matrix_1.4-0                vegan_2.6-0                
##  [85] permute_0.9-7               vctrs_0.4.1                
##  [87] pillar_1.7.0                lifecycle_1.0.1            
##  [89] jquerylib_0.1.4             BiocNeighbors_1.12.0       
##  [91] bitops_1.0-7                irlba_2.3.5                
##  [93] patchwork_1.1.1             R6_2.5.1                   
##  [95] bookdown_0.24               vipor_0.4.5                
##  [97] MASS_7.3-55                 assertthat_0.2.1           
##  [99] withr_2.5.0                 GenomeInfoDbData_1.2.7     
## [101] mgcv_1.8-38                 parallel_4.1.2             
## [103] ggfun_0.0.5                 beachmat_2.10.0            
## [105] tidyr_1.2.0                 rmarkdown_2.11.3           
## [107] DelayedMatrixStats_1.16.0   carData_3.0-5              
## [109] ggpubr_0.4.0                ggnewscale_0.4.6           
## [111] ggforce_0.3.3               ggbeeswarm_0.6.0</code></pre>
</div>
<div id="bibliography" class="section level2" number="13.6">
<h2><span class="header-section-number">13.6</span> Bibliography</h2>
<p>Robert L. Thorndike. 1953. “Who Belongs in the Family?”. Psychometrika. 18 (4): 267–276. <a href="doi:10.1007/BF02289263" class="uri">doi:10.1007/BF02289263</a></p>
<p>Peter J. Rousseeuw. 1987. “Silhouettes: a Graphical Aid to the Interpretation and Validation of Cluster Analysis”. Computational and Applied Mathematics. 20: 53–65. <a href="doi:10.1016/0377-0427(87)90125-7" class="uri">doi:10.1016/0377-0427(87)90125-7</a>.</p>
<p>Robert Tibshirani, Guenther Walther, and Trevor Hastie. 2002. Estimating the number of clusters in a data set via the gap statistic method. (63): 411-423. <a href="doi:10.1111/1467-9868.00293" class="uri">doi:10.1111/1467-9868.00293</a>.</p>
<p>Susan Holmes and Joey McMurdie. 2015. Reproducible Research: Enterotype Example. <a href="http://statweb.stanford.edu/~susan/papers/EnterotypeRR.html" class="uri">http://statweb.stanford.edu/~susan/papers/EnterotypeRR.html</a>.</p>
<p>Daniel B. DiGiulio et al. 2015. Temporal and spatial variation of the human microbiota during pregnancy. (112): 11060–11065. <a href="doi:10.1073/pnas.1502875112" class="uri">doi:10.1073/pnas.1502875112</a></p>
<p>Arumugam, M., et al. (2011). Enterotypes of the human gut microbiome.</p>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="study-material.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="exercise-solutions.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": true,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["oulu2022_material.pdf"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
